### PR TITLE
TransactionsHandle propagation commands should not adhere to caching

### LINKED ISSUE #12033: TransactionsHandle propagation commands should not adhere to caching
### Describe the feature

currently most of manual propagation commands 

https://github.com/paradigmxyz/reth/blob/d50da7fcd65da3c37ad77b7b5f645f12ffc055b6/crates/net/network/src/transactions/mod.rs#L879-L894

use the regular broadcast impls, which also check caches:

https://github.com/paradigmxyz/reth/blob/d50da7fcd65da3c37ad77b7b5f645f12ffc055b6/crates/net/network/src/transactions/mod.rs#L592-L596

https://github.com/paradigmxyz/reth/blob/d50da7fcd65da3c37ad77b7b5f645f12ffc055b6/crates/net/network/src/transactions/mod.rs#L530-L532

this makes it impossible to force broadcasts manually.
This can be fixed with a simple additional enum argument that determines whether we should filter seen transactions or not

## TODO
* introduce this enum 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Parikalp-Bhardwaj**
Closes #12033

**Description:**
This pull request addresses the issue where manual propagation commands in `TransactionsHandle` were adhering to caching, preventing forced broadcasts. The following changes have been made:

- **Added `PropgateKind` Enum:**
  Introduced a `PropagateKind` enum with variants `Basic` and `Forced` to control caching behavior during transaction propagation.

- **Modified Propagation Methods:**
  Updated methods such as `propagate_full_transactions_to_peer`, `propagate_hashes_to`, and `propagate_transactions` to accept a `PropgateKind` parameter.

- **Updated Manual Propagation Commands:**
  In the `on_command` method, manual propagation commands now use `PropgateKind::Basic` to bypass the cache, allowing forced broadcasts.

- **Ensured Regular Propagation Respects Cache:**
  Regular propagation commands continue to use `PropgateKind::Forced`, maintaining the caching mechanism to optimize network usage.

- **Updated Tests:**
  Modified existing tests and added new ones to cover scenarios with both cache policies, ensuring that the caching behavior is as expected.

@mattsse  Please review the changes and let me know if any adjustments are needed.

### CODE REVIEW

File: crates/net/network/src/transactions/mod.rs [hunk 1]
```diff
@@ -454,9 +462,16 @@ where
...
-                    builder.push(tx);
+                    CachePolicy::IgnoreCache => {
+                        builder.push(tx);
+                    }
+                    CachePolicy::RespectCache => {
+                        if !peer.seen_transactions.contains(&tx.hash()) {
+                            builder.push(tx);
+                        }
+                    }
                 }
```
Comment (mattsse) [inline crates/net/network/src/transactions/mod.rs:475]: let's convert this into an if by adding a `fn is_forced` instead

File: crates/net/network/src/transactions/mod.rs [hunk 2]
```diff
@@ -260,6 +260,12 @@ pub struct TransactionsManager<Pool> {
     metrics: TransactionsManagerMetrics,
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum CachePolicy {
+    IgnoreCache,
+    RespectCache,
+}
```
Comment (mattsse) [inline crates/net/network/src/transactions/mod.rs:267]: I'd like to rename this slightly to something like

```rust
enum PropgateKind {
  /// Default propgation mode, filters out txs that we already sent or received
  Basic,
  /// Always propagate, even if we already sent or received the txs.
  Forced,
}

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, functionally this is correct.

only have style nits

**Review by mattsse — CHANGES_REQUESTED** [review]
ah, I see we already have an enum that is used with that name.

we need a new enum just for this, with the two variants. this should not change the existing enum

**Comment by Parikalp-Bhardwaj** [comment]
@mattsse done, Created a new enum `PropgateKind` with the two variants.

**Review by mattsse — APPROVED** [review]

**Pull request merged**